docs/proposals/cross-tenant-query-federation.md

```diff
@@ -0,0 +1,99 @@
+---
+title: "Cross-Tenant Query Federation"
+linkTitle: "Cross-Tenant Query Federation"
+weight: 1
+slug: "cross-tenant-query-federation"
+---
+
+- Author: [Christian Simon](https://github.com/simonswine)
+- Date: October 2020
+- Status: Proposed
+
+## Overview
+
+This document aims to describe how to implement the ability to allow queries to cover data from more than a single Cortex tenant.
+
+## Reasoning
+
+Adopting a tenancy model within an organization with each tenant representing a department comes with the disadvantage that it will prevent queries from spanning multiple departments. This proposal tries to overcome those limitations.
+
+## Alternatives considered
+
+In theory PromQL API clients could be aggregating/correlating query results from multiple tenants. For example Grafana could be used with multiple data sources and a cross tenant query could be achieved through using [Transformations][grafana_transformation].
+
+[grafana_transformation]: https://grafana.com/docs/grafana/latest/panels/transformations/
+
+## Challenges
+
+### Aggregate data without overlaps
+
+#### Challenge
+
+The series in different tenants might have exactly the same labels and hence potentially collide which each other.
+
+#### Proposal
+
+In order to be able to always identify the tenant correctly, queries using multiple tenants should inject a tenant label and its value containing the tenant ID into the results. An potentially existing label with the same name should be stored in a label with the prefix `original_`. This also should be done recursively until no more conflicts arise.
+
+The label name used should be configurable through the flag `-experimental.querier.tenant-federation.label` with a default value of `__tenant__`.
+
+Label selectors containing the tenant label should behave like any other labels. This can be achieved by selecting the tenants used in a multi tenant query.
+
+### Exposure of feature to the user
+
+#### Challenge
+
+The tenant ID is currently read from the `X-Scope-OrgID` HTTP header. There is no limitation on character classes defined, that can be used as part of the tenant ID.
+
+#### Proposal
+
+For the query path a user should be able to specify a `X-Scope-OrgID` per tenant ID and the information should be propagating as a string slice throughout out the system until it reaches the querier. The `Queryable` instances in the querier component, are wrapped by a `mergeQueryable`, which will aggregate the results from a `Queryable` per tenant and hence treated by the downstream components as a single tenant query.
```

Comment:
> For the query path a user should be able to specify a `X-Scope-OrgID` per tenant ID

If you mean specifying the HTTP header multiple times, be aware of the HTTP specifications:
https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.2

In particular:
> Multiple message-header fields with the same field-name MAY be present in a message if and only if the entire field-value for that header field is defined as a comma-separated list [i.e., #(values)]. It MUST be possible to combine the multiple header fields into one "field-name: field-value" pair, without changing the semantics of the message, by appending each subsequent field-value to the first, each separated by a comma. The order in which header fields with the same field-name are received is therefore significant to the interpretation of the combined field value, and thus a proxy MUST NOT change the order of these field values when a message is forwarded.

Reply:
Very good point. I wasn't aware that the requirement is that strict.

After doing some research, I have found this implementation of splitting a header value without reducing the character set:

https://github.com/golang/gddo/blob/master/httputil/header/header.go#L86

I think that mechanism should work for us too

Reply:
Thanks for sharing. I took a quick look. So looks like it has to be comma separated, but values can (not must) be quoted to be able to use commas within the single value. Correct?

Reply:
Yes that's how I read the BNF as part of the RFC (https://tools.ietf.org/html/rfc2616#section-2.2).

It's further more possible to include `"` by using `\"`.

Reply:
Instead of using a third party library to handle mutli-part headers, could we just assume each tenant ID header value will be URL encoded using `url.Path(Escape|Unescape)`? This will not affect anyone not currently encoding URL encoding their tenant ID but should accommodate this use case. The only bug this might introduce is if a user currently has a tenant ID that includes a `%` character, it may trigger an error unless they ensure the `%` is URL escaped. However, I can't imagine a `%` character in a tenant ID is something that is very common.

See more here: https://play.golang.org/p/YjS5fWT8K-y

Reply:
(Joining bike-shedding party) I'd suggest using `base64/` prefix and using base64-encoded comma-separated list of tenants after it (eg. `base64/aGVsbG8sd29ybGQ=` for "hello,world"). If header value doesn't have this prefix, we wouldn't interpret it. This should work since we already don't support `/` in tenants.

I wouldn't recommend using `url.PathEscape` since it has different purpose (namely escaping strings in URL paths as name suggests).

(That said, other changes related to propagating this through Cortex will be more "fun").

Reply:
> (Joining bike-shedding party) I'd suggest using `base64/` prefix and using base64-encoded comma-separated list of tenants after it (eg. `base64/aGVsbG8sd29ybGQ=` for "hello,world"). If header value doesn't have this prefix, we wouldn't interpret it. This should work since we already don't support `/` in tenants.

I think that could be a good way forward. We still need to think about escaping of a comma for example. I would suggest we just use `json.Marshal([]string{"tenant-,",  "tenant-b"})` and then encode it using base64 as suggested.

> (That said, other changes related to propagating this through Cortex will be more "fun").

I think it probably entails having our own implementation of `weaveworks/common/user` and then make it other libraries in `weaveworks/common` extensible in the sense to bring your own implementation of it using an interface.

Reply:
I think that basically both options could work fairly well. I think we should make a decision about this soon, so I can update the proposal and prepare the work that is required to make this happen

A :rocket: ) Use a single header, if the header starts with prefix `base64/`, the remainder of the value is a base64 encoded ([URL variant]) JSON list.

[URL variant]: https://github.com/golang/go/blob/d4c1ad882973e407ff85b977f4ce5b9435451190/src/encoding/base64/base64.go#L103

B :tada:) Support multiple header values and as RFC2616 4.2 demands a comma separated list as discussed above.

I think I am leaning towards option (A :rocket: ) as it will allow us to extend it in the future and en-/decoding will be super straightforward to implement. It comes with the advantage of limited readability of the header itself

/cc opinions/votes welcome @jtlisi @pracucci @tomwilkie @gouthamve

Reply:
I've the feeling we're making it too complicated. Some thoughts:
1. Comma `,` is not an officially supported character for tenant ID (see https://cortexmetrics.io/docs/guides/limitations/#tenant-id-naming).
2. We should try to make it easy to use (think about engineers running a `curl -H "X-ScopeOrgID: ..."` command. Doing the base64 encoding of a json is very annoying on the command line.
3. We could conditionally enable the parsing of `,` in the `X-Scope-OrgID` only when query federation is enabled. This would make sure we don't introduce any breaking change in Cortex tenant ID if we just split the header value on comma.

Reply:
Even better, we could use `|` instead of `,`. The pipe in the tenant ID should be even less common and it's considered a [character to avoid](https://docs.aws.amazon.com/AmazonS3/latest/dev/UsingMetadata.html) by S3.

Reply:
>Even better, we could use | instead of ,. The pipe in the tenant ID should be even less common and it's considered a character to avoid by S3 (why S3? because if a character is unsupported by S3 or discouraged by S3, we could assume none is using it in Cortex).

I'm 100% behind this approach.

Reply:
Thanks everyone for you input. I agree @pracucci 's proposal is the right sized solution for this. I have updated the proposal and mark this as resolved.